wallet/chainntfns.go

```diff
@@ -338,6 +342,82 @@ func (w *Wallet) handleReorganizing(oldHash *chainhash.Hash, oldHeight int64,
 	chainClient.SetReorganizingState(true, *newHash)
 }
 
+// evaluateStakePoolTicket evaluates a stake pool ticket to see if it's
+// acceptable to the stake pool. The ticket must pay out to the stake
+// pool cold wallet, and must have a sufficient fee.
+func (w *Wallet) evaluateStakePoolTicket(rec *wtxmgr.TxRecord,
+	block *wtxmgr.BlockMeta, poolUser dcrutil.Address) (bool, error) {
+	tx := rec.MsgTx
+
+	// Check the first commitment output (txOuts[1])
+	// and ensure that the address found there exists
+	// in the list of approved addresses. Also ensure
+	// that the fee exists and is of the amount
+	// requested by the pool.
+	commitmentOut := tx.TxOut[1]
+	commitAddr, err := stake.AddrFromSStxPkScrCommitment(
+		commitmentOut.PkScript, w.chainParams)
+	if err != nil {
+		return false, fmt.Errorf("Failed to parse commit out addr: %s",
+			err.Error())
+	}
+
+	// Extract the fee from the ticket.
+	in := dcrutil.Amount(0)
+	for i := range tx.TxOut {
+		if i%2 != 0 {
+			commitAmt, err := stake.AmountFromSStxPkScrCommitment(
+				tx.TxOut[i].PkScript)
+			if err != nil {
+				return false, fmt.Errorf("Failed to parse commit "+
+					"out amt for commit in vout %v: %s", i, err.Error())
+			}
+			in += dcrutil.Amount(commitAmt)
+		}
+	}
+	out := dcrutil.Amount(0)
+	for i := range tx.TxOut {
+		out += dcrutil.Amount(tx.TxOut[i].Value)
+	}
+	fees := out - in
+
+	_, exists := w.stakePoolAddrs[commitAddr.EncodeAddress()]
+	if exists {
+		commitAmt, err := stake.AmountFromSStxPkScrCommitment(
+			commitmentOut.PkScript)
+		if err != nil {
+			return false, fmt.Errorf("failed to parse commit "+
+				"out amt: %s", err.Error())
+		}
+
+		// Calculate the fee required based on the current
+		// height and the required amount from the pool.
+		feeNeeded := txrules.StakePoolTicketFee(dcrutil.Amount(
+			tx.TxOut[0].Value), fees, block.Height, w.PoolFees(),
+			w.ChainParams())
+		if commitAmt < feeNeeded {
+			log.Warnf("User %s submitted ticket %v which "+
+				"has less fees than are required to use this "+
+				"stake pool and is being skipped (required: %v"+
+				", found %v)", commitAddr.EncodeAddress(),
+				tx.TxSha(), feeNeeded, commitAmt)
+
+			// Reject the entire transaction if it didn't
+			// pay the pool server fees.
+			return false, nil
+		}
+	} else {
+		log.Warnf("Unknown pool commitment address %s for ticket %v",
+			commitAddr.EncodeAddress(), tx.TxSha())
+		return false, nil
+	}
+
+	log.Debugf("Accepted valid stake pool ticket %v commiting %v in fees",
+		tx.TxSha(), dcrutil.Amount(tx.TxOut[0].Value).ToCoin())
```

Comment:
Don't use ToCoin.  This will return a float64 and is not guaranteed to be logged with the correct precision.